### ISSUE TITLE
Detect incorrect or questionable use of TestFixture inheritance.

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit.analyzers#840]
When you derive a TestFixture from another TestFixture, NUnit generally handles the situation as follows:

1. It runs the base class tests against an instance of the base class.
2. It runs the base class tests a second time against an instance of the derived class.

Note that (2) is repeated for **every** derived class

This is almost never what the user intended. Normally, what is expected is for the base class tests to only be run once for each derived class, not n + 1 times. That behavior is correctly achieved by using the "Abstract Test Fixture" pattern. If the base class is abstract, then NUnit will not attempt to construct it or run tests against it.

A special case of this problem arises if the TestFixture is parameterized and  no paramterized instance of `[TestFixture()]` In that case, rather than running the contained tests n + 1 times, an error will arise indicating that no valid constructor can be found. 

Both situations (with or without an error arising) should probably be dealt with.

In reviewing issues and questions, I have occasionally found a user who actually __wanted__ the tests to run twice (or n + 1 times). I believe most of the examples used logic in the base class to make it behave differently when executed directly by examining the Type of the instance. This seems pathological to me and I would not recommend supporting it.

**Comment by manfred-brands** [comment]
As the `[TestFixture]` attribute is optional, this hold for any class with a `[Test]` and other equivalents like `[TestCaseSource]`.

A test fixture class should therefore be either `sealed` or `abstract` to prevent this.

In the below example, the `Parent` tests are executed standalone, as part of the `Child` class and as part of the `GrandChild` class:

![Image](https://github.com/user-attachments/assets/8b747c8b-2af8-46c3-8b78-c619dd92b3ef)

```csharp
namespace Tests
{
    internal class Parent
    {
        private static readonly string[] Names = ["Hello", "World"];

        [TestCaseSource(nameof(Names))]
        public void ParentRun(string name)
        {
            Assert.Pass($"Run from {GetType().Name} using {name}");
        }
    }

    internal class Child : Parent
    {
        [Test]
        public void ChildRun()
        {
            Assert.Pass($"Run from {GetType().Name}");
        }
    }

    internal class GrandChild : Child
    {
        [Test]
        public void GrandChildRun()
        {
            Assert.Pass($"Run from {GetType().Name}");
        }
    }
}
```

**Comment by manfred-brands** [comment]
> A test fixture class should therefore be either sealed or abstract to prevent this.

In my company we had the `abstract` or `sealed` (and `internal`) rule for years, for this exact reason because I noticed a base class executing standalone.

But I suspect that this rule would fire up for a lot of tests, including in _nunit_ itself that are not `sealed`, but also nothing derives from it. A single codefix can solve it in one go so it is a one-off. 
Maybe we can detect if a class is not derived from and only complain about the ones that are.

**Comment by CharliePoole** [comment]
@manfred-brands 

I like your rule. It takes a problem that would otherwise require a global analysis and converts it to one that can be detected by looking at a single class definition. A refinement would be to not require `sealed` if the class contained no tests.

Slightly off-topic but related: Does the analyzer detect the presence of both `[TestFixture]` and `[SetUpFixture]` on the same class? That's an error I've run into a number of times on user-reported issues and on SO. Also, having tests in a SetUpFixture, which amounts to the same thing.

**Comment by manfred-brands** [comment]
@CharliePoole 

> t takes a problem that would otherwise require a global analysis and converts it to one that can be detected by looking at a single class definition
It unfortunately has false positives as it would seal base classes unless they were already explicitly marked `abstract` which is likely to upset too many user that haven't been consistent defining classes as either `abstract` or `sealed` already.

> A refinement would be to not require sealed if the class contained no tests.
It then should be `abstract`.

> Slightly off-topic but related: Does the analyzer detect the presence of both [TestFixture] and [SetUpFixture] on the same class?
There are no rules at all for `SetUpFixture` classes.
